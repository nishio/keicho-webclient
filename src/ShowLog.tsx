import React, { useEffect } from "react";
import { useGlobal } from "reactn";
import { ChatLine } from "./ChatContents";
import { loadLogs } from "./loadLogs";
import { ButtonAppBarForShowLog } from "./Menu";
import { RegroupDialog } from "./RegroupDialog";
import { ScrapboxDialog } from "./ScrapboxDialog";
import { TalkListDialog } from "./TalkListDialog";
export let loadLogsPromise: Promise<unknown>;
export const ShowLog = (props: { talk: string }) => {
  const [logs] = useGlobal("logs");

  useEffect(() => {
    loadLogsPromise = loadLogs(props.talk);
  }, [props.talk]);
  return (
    <>
      <ButtonAppBarForShowLog />
      <ChatLine logs={logs}></ChatLine>
      <TalkListDialog />
      <RegroupDialog />
      <ScrapboxDialog />
    </>
  );
};
