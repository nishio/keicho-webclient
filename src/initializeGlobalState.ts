import { setGlobal } from "reactn";
import { INITIAL_LOGS, PRESET_LOGS, USE_PRESET } from "./PRESET_MESSAGES";

const INITIAL_GLOBAL_STATE = {
  TalkID: "",
  logs: USE_PRESET ? PRESET_LOGS : INITIAL_LOGS,
  lastKeywords: [] as string[],
  otherKeywords: [] as string[],
  canInput: true,
  previousTalkID: "",
};

export const initializeGlobalState = () => {
  setGlobal(INITIAL_GLOBAL_STATE);
};

type TYPE_GLOBAL_STATE = typeof INITIAL_GLOBAL_STATE;

declare module "reactn/default" {
  export interface State extends TYPE_GLOBAL_STATE {}
}
