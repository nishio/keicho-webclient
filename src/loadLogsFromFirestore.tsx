import db from "./FirestoreIO";
import { Document } from "./loadLogs";

export const loadLogsFromFirestore = (talk: string) => {
  return db
    .collection("yagokoro_env")
    .doc(talk)
    .get()
    .then((doc: Document) => {
      if (doc.exists) {
        const data = doc.data();
        if (data === undefined) {
          throw new TypeError("doc is undefined");
        }
        return JSON.parse(data.json);
      }
    });
};
