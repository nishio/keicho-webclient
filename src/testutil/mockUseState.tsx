import React, { Dispatch } from "react";
import { act } from "@testing-library/react";
import { useState as originalUseState } from "react";

export const mockUseState = () => {
  jest.spyOn(React, "useState").mockImplementation((arg?: unknown): [
    unknown,
    Dispatch<unknown>
  ] => {
    const [s, setS] = originalUseState(arg);
    return [
      s,
      (arg: unknown) => {
        act(() => {
          setS(arg);
        });
      },
    ];
  });
};
