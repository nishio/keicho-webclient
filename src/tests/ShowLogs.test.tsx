import React from "react";
import { unmountComponentAtNode } from "react-dom";
import { initializeGlobalState } from "../initializeGlobalState";

import { act, fireEvent, render, screen } from "@testing-library/react";
import * as loadLogsFromFirestoreModule from "../loadLogsFromFirestore";
import * as MockTalkObject from "../talkObject.json";
import { loadLogsPromise, ShowLog } from "../ShowLog";
import { mockUseState } from "../testutil/mockUseState";

jest.spyOn(window, "scrollTo").mockImplementation(() => {});

let container: HTMLDivElement;

beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
  act(() => {
    initializeGlobalState();
  });
  mockUseState();
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
});

test("show log", async () => {
  const m = jest
    .spyOn(loadLogsFromFirestoreModule, "loadLogsFromFirestore")
    .mockResolvedValue(MockTalkObject);

  // no talkID needed beacuse loadLogsFromFirestore is mocked
  render(<ShowLog talk="" />);
  await loadLogsPromise;
  expect(m).toHaveBeenCalled();
  expect(screen.getByText("üôÅ")).toBeTruthy();
  fireEvent.click(screen.getByLabelText("menu"));
  fireEvent.click(screen.getByText("Export for Regroup"));
  m.mockRestore();
});
