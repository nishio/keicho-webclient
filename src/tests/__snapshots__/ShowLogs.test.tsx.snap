// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show log 1`] = `
<textarea
  data-testid="textarea-export-for-regroup"
  id="multiline"
  rows="1"
  style="width: 100%; height: -4px; overflow: hidden;"
>
  あなたはこの会話で何が起きて欲しいですか？
終わりのデザインの能動学習を実装したんだけどイマイチなので相談したい

その「能動学習」は、どんな「能動学習」ですか？
能動学習
機械学習の一手法で、システムが自信のないところを人間に質問し、それに人間が答えることで効率的に教師データを集める手法

その「手法」は、どんな「手法」ですか？
手法
ようはAIを効率よく賢くする手法

その「手法」について、他に何かありますか？
手法
システム的には質問と回答のサイクルは回り始めてるのだけど、現在のAIが頭悪過ぎてあんまり効率が良くない

その「AI」は、どんな「AI」ですか？
AI
ここでやろうとしてるのは、対話の流れを見て適切なところで「ここは終わるのに良いところか？」と質問するもの


&gt; ようはAIを効率よく賢くする手法

この「手法」は、どのあたりにありますか？
手法
「動きの抽出」の時に実装したのを再利用した


&gt; ようはAIを効率よく賢くする手法

この「手法」は、どこにありますか？
手法
🙁


&gt; ようはAIを効率よく賢くする手法

この「手法」は、何のようですか？
手法
子供が興味を持ったことを質問してくるみたいな感じ

その「興味」は、どんな「興味」ですか？
興味
YesかNoかを識別したい時に、ボーダーライン上の事例がどう識別されるべきかを確認する

その「ボーダーライン上」は、どんな「ボーダーライン上」ですか？
ボーダーライン上
Yesである確率が50%に近いもの

その「Yes」は、どんな「Yes」ですか？
Yes
ここでは「ここで終わりの質問をするべき」という判断

その「質問」は、どんな「質問」ですか？
質問
「ここは終わるのに良いところか？」

あなたが価値を感じているのは何ですか？
良いタイミングでその質問をしてくれて「はい」「じゃあ終わりにしましょう」って感じで終わると、ダラダラ対話が続くより気持ちが良い

その「質問」について、他に何かありますか？
質問
どういうタイミングでするべきかを僕もまだあまり言語化できてないんだよな、だから機械学習でやろうとした

その「機械学習」は、どんな「機械学習」ですか？
機械学習
人間が条件式を記述できない時に、データを集めることによって機械が条件を見出す仕組み

その「人間」は、どんな「人間」ですか？
人間
僕です

その「僕」は、どんな「僕」ですか？
僕
このシステムを実装しようとしているが、どういう時に終わりの質問を出すべきかまだ明確に言語化できてない

その「質問」は、どのあたりにありますか？
質問
だいたいある程度対話が発展して落ち着いて来たあたりにあるべき


&gt; このシステムを実装しようとしているが、どういう時に終わりの質問を出すべきかまだ明確に言語化できてない

この「質問」は、どこにありますか？
質問
質問に答えることでユーザが何かを獲得した後で、それ以上質問を続けても得るものが少なそうな時に出す

あなたはその「質問」をどう知るのですか？
質問
僕自身が人間を対面でコーチングする場合には表情とか身振りとかテンションから判断するのだけど、このシステムはチャットなので情報が取りにくいね

その「システム」は、どんな「システム」ですか？
システム
対話によってユーザの内面の言語化を促すシステム

その「システム」について、他に何かありますか？
システム
言語化を促すわけなので、何か言語化されたら一旦は役目が完了しているわけだ

その「言語化」は、どんな「言語化」ですか？
言語化
このシステムを使う前にユーザがあるテーマについて書き物をして、システムを使った後にも書き物をして、この二つに差があるならその差がこのシステムによってもたらされた言語化支援効果

その「システム」は、何のようですか？
システム
聞き出すことを中心に据えるタイプのコーチングはこんな感じ

その「感じ」は、どんな「感じ」ですか？
感じ
システムが答えを与えるのではなく、質問を与えることでユーザの中で結晶化が起こることを期待する、というあたりが同じ

その「質問」は、どこから来るのですか？
質問
クリーンランゲージとかシンボリックモデリングと呼ばれてる手法から取った

あなたはその「手法」をどう知るのですか？
手法
確かに、参考にした書籍などに終わりの質問をする際の判断基準が書いてあるかもしれないな

その「質問」に関して、何が起こるとよいですか？
質問
質問をどのタイミングでするかに関する判断材料がなんであるかに関してはもっと言語化が進むといい

それには「質問」に何が起こる必要がありますか？
僕自身がこの質問をするべきタイミングについてもっと詳しくなるべき

「質問」に関して、それができますか？
参考書籍を僕が読み返すことでできそう


&gt; 僕自身がこの質問をするべきタイミングについてもっと詳しくなるべき

この「質問」は、何のようですか？
質問
(ここで「ここは終わるのに良いところですか？」「はい」とやって終わりたい

</textarea>
`;
